Update Workload API spec with JWT-SVID Profile (#144)
spiffe/spiffe: standards/SPIFFE_Workload_API.md

4.1. [Identifying the Caller](#41-identifying-the-caller)
4.2. [Connection Lifetime](#42-connection-lifetime)
4.3. [Stream Responses](#43-stream-responses)
4.4. [Default Values and Redacted Information](#44-default-values-and-redacted-information)
4.5. [Mandatory Fields](#45-mandatory-fields)
4.6. [Federated Bundles](#46-federated-bundles)
5\. [X.509-SVID Profile](#5-x509-svid-profile)
5.1. [Profile Definition](#51-profile-definition)
5.2. [Profile RPCs](#52-profile-rpcs)
5.3. [Default Identity](#53-default-identity)
6\. [JWT-SVID Profile](#6-jwt-svid-profile)
6.1. [Profile Definition](#61-profile-definition)
6.2. [Profile RPCs](#62-profile-rpcs)
6.3. [JWT-SVID Validation](#63-jwt-svid-validation)

## 1. Introduction

The SPIFFE Workload API is an API which provides information and services that enable workloads, or compute processes, to leverage SPIFFE identities and SPIFFE-based authentication systems. It is served by the [SPIFFE Workload Endpoint](SPIFFE_Workload_Endpoint.md), and comprises a number of services, or *profiles*.

Currently, there are two profiles:

- [X.509-SVID Profile](#5-x509-svid-profile)
- [JWT-SVID Profile](#6-jwt-svid-profile)

> Both profiles are mandatory and MUST be implemented. However, one or the other MAY be administratively disabled.

Reviewer: I'm not sure how to reconcile these two sentences. If a profile may be disabled, then why _must_ the profile be implemented?

Reviewer 2: I think this is down to lack of specificity _on whom and where_ we are imposing specification requirements, i.e. _implementers_ are required to implement both to claim their _implementation_ meets the spec.  _Operators_ are not required to enable both on any specific _deployment_.

Author: I added some clarifying language.

Future versions of this specification may introduce additional profiles or make one or more profiles optional.

## 2. Extensibility

> The SPIFFE Workload API MUST NOT be extended beyond this specification. Implementers wishing to provide extended functionality may do so by introducing new gRPC services, according to the [extensibility method](SPIFFE_Workload_Endpoint.md#7-extensibility-and-services-rendered) outlined in the SPIFFE Workload Endpoint specification.

Reviewer: Out of curiosity, why must the API not be extended? Is is to preserve cross-implementation compatibility? (Might want to explain the reasoning here in the text.)

Author: Good catch! This is text is unrelated to the JWT-SVID profile addition (or subsequent rejiggering of the text). Can I suggest we defer edits here to a subsequent PR?

Author: Talked IRL; punting to another PR.

The exact timing of server response messages is implementation-specific, and SHOULD be dictated by events which change the response, such as an SVID rotation, a CRL update, etc. Receiving a request message from the client MUST be considered a response-generating event. In other words, the first response message of the server response stream (on a connection-by-connection basis) MUST be sent as soon as possible, without delay.

Finally, implementers of SPIFFE Workload API servers should be careful about pushing updated response messages *too* rapidly. Some software may reload automatically upon receiving new information, potentially causing a period of unavailability should all instances reload at once. As a result, implementers may introduce some splay/jitter in the transmission of widespread updates.

For additional clarity, please see [Appendix A](#appendix-a.-sample-implementation-state-machines) for sample implementation state machines.

### 4.4 Default Values and Redacted Information

> SPIFFE Workload API clients may at times encounter fields in the response message that have a default value, or may notice that information included in a previous response is not included in the latest response. For instance, a client may encounter a default value in the `federated_bundles` field after having previously received a federated bundle.

Reviewer: Consider rephrasing the first sentence to something like:
SPIFFE Workload API response messages are complete updates to previously sent response messages. When a response message contains fields which are set to default or empty values, clients MUST interpret the values of those fields to have been set to their default or empty values; previously received, non-default or non-empty values MUST NOT be retained by a client after receiving a default or empty value for the fields.

Author: Awesome. I took this verbatim and also tweaked following sentence to fill out the example a little more.

### 4.5 Mandatory Fields

Messages exchanged for the profile RPCs are comprised of both mandatory and optional fields. Servers receiving a message in which a mandatory field has a default value SHOULD respond with the "InvalidArgument" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information). Clients receiving a message in which a mandatory field has a default value SHOULD report an error and discard the message.

### 4.6 Federated Bundles

Various RPCs defined in this specification can return trust bundles for foreign trust domains. Inclusion of foreign bundles enables workloads to communicate *across* trust domains, and is the primary mechanism through which federation is enabled. A bundle representing a foreign trust domain is known as a *Federated Bundle*.

> When authenticating a client from a foreign trust domain, the authenticator chooses the bundle representing the client’s presented trust domain for validation. Similarly, when authenticating a server, the client uses the bundle representing the server’s trust domain. If no matching bundle is present for the SVID in use, then the peer is untrusted. This approach is required in order to account for the lack of widespread support for SAN URI Name Constraints in common X.509 libraries. Please see [Section 4.2](X509-SVID.md#42-name-constraints) of the X509-SVID specification for more information.

Reviewer: Would the first sentence be just as true if written:
When authenticating a client, the authenticator ...

What caught my attention is that the next sentence does not contain the phrase "from a foreign trust domain".

Author: I think so. Fixed.

> When authenticating a client from a foreign trust domain, the authenticator chooses the bundle representing the client’s presented trust domain for validation. Similarly, when authenticating a server, the client uses the bundle representing the server’s trust domain. If no matching bundle is present for the SVID in use, then the peer is untrusted. This approach is required in order to account for the lack of widespread support for SAN URI Name Constraints in common X.509 libraries. Please see [Section 4.2](X509-SVID.md#42-name-constraints) of the X509-SVID specification for more information.

Reviewer: I recommend either explaining the relevance of support for SAN URI Name Constraints or remove reference it from this paragraph.

My preference would be to remove the reference since I think it's a distraction here, even if the relevance was explained.

Author: I'm OK removing the reference. It is part of the original text however. @spikecurtis, @evan2645 are you also comfortable removing this reference?

Author: Talked IRL. Punting this to a separate PR.

## 5. X.509-SVID Profile

The X.509-SVID Profile of the SPIFFE Workload API provides a set of gRPC methods which can be used by workloads to retrieve [X.509-SVIDs](X509-SVID.md) and their related trust bundles. This profile outlines the signature of these methods, as well as related client and server behavior.

### 5.1 Profile Definition

The X.509-SVID Profile RPCs and associated messages are defined below. For the complete Workload API service definition, see [workloadapi.proto](workloadapi.proto).

```protobuf
service SpiffeWorkloadAPI {
    /////////////////////////////////////////////////////////////////////////
    // X509-SVID Profile
    /////////////////////////////////////////////////////////////////////////

    // Fetch X.509-SVIDs for all SPIFFE identities the workload is entitled to,
    // as well as related information like trust bundles and CRLs. As this

>     // information changes, subsequent messages will be sent.

Reviewer: Instead of "sent", consider "streamed from the server".

Author: Updated all instances.

    rpc FetchX509SVID(X509SVIDRequest) returns (stream X509SVIDResponse);

    // Fetch trust bundles and CRLs. Useful for clients that only need to
    // validate SVIDs without obtaining an SVID for themself. As this

>     // information changes, subsequent messages will be sent.

Reviewer: Same comment as above, consider s/sent/streamed from the server/.

    rpc FetchX509Bundles(X509BundlesRequest) returns (stream X509BundlesResponse);

    // ... RPCS for other profiles ...
}

// The X509SVIDRequest message conveys parameters for requesting an X.509-SVID.

> // There are currently no such parameters.

Reviewer: s/such/request/

message X509SVIDRequest {  }

// The X509SVIDResponse message carries X.509-SVIDs and related information,

> // including a global CRL and list of bundles the workload is federated with.

Reviewer: Consider replacing "list of bundles the workload is federated with" with
"a list of foreign trust bundles which the workload may use to federate with".

1) s/list/a list/
2) are the bundles necessarily foreign bundles? I'm not sure, please clarify.
3) does the presence of the bundles imply that the workload "is federated with" them? or merely helps enable that federation capability?

Author: I took a stab at rewording this.

message X509SVIDResponse {
    // Required. A list of X509SVID messages, each of which includes a single
    // X.509-SVID, its private key, and the bundle for the trust domain.
    repeated X509SVID svids = 1;

    // Optional. An ASN.1 DER encoded CRL.
    repeated bytes crl = 2;

    // Optional. CA certificate bundles belonging to foreign trust domains that the
    // workload should trust, keyed by the SPIFFE ID of the foreign
    // domain. Bundles are ASN.1 DER encoded.
    map<string, bytes> federated_bundles = 3;
}

// The X509SVID message carries a single SVID and all associated
// information, including X.509 bundle for the trust domain.
message X509SVID {
    // Required. The SPIFFE ID of the SVID in this entry
    string spiffe_id = 1;

>     // Required. ASN.1 DER encoded certificate chain. MAY include

Reviewer: While it's clear that the leaf cert must be at the head of the list, does the order of the intermediates (following the leaf) matter?

Author: Not in my experience, no.

    // intermediates, the leaf certificate (or SVID itself) MUST come first.
    bytes x509_svid = 2;

    // Required. ASN.1 DER encoded PKCS#8 private key. MUST be unencrypted.
    bytes x509_svid_key = 3;

    // Required. ASN.1 DER encoded X.509 bundle for the trust domain.
    bytes bundle = 4;
}

// The X509BundlesRequest message conveys parameters for requesting X.509
// bundles. There are currently no such parameters.
message X509BundlesRequest {
}

// The X509BundlesResponse message carries a global CRL and a
// map of trust bundles the workload should trust.
message X509BundlesResponse {
    // Optional. ASN.1 DER encoded certificate revocation list.
    repeated bytes crl = 1;

    // Required. CA certificate bundles belonging to trust domains that the
    // workload should trust, keyed by the SPIFFE ID of the trust
    // domain. Bundles are ASN.1 DER encoded.
    map<string, bytes> bundles = 2;
}

message JWTSVIDRequest {

>     // Required. The audience the workload intends to authenticate against.

Reviewer: Since the field is marked `repeated`, should "audience" be "audience(s)" ?

Author: :+1:

    repeated string audience = 1;

    // Optional. The requested SPIFFE ID for the JWT-SVID. If unset, JWT-SVIDs

>     // for all identities the workload is entitled to are returned.

Reviewer: Since this comment belongs to the JWTSVIDRequest (not the Response), I think we should describe the semantics of the request as opposed to expected values in the response. Consider rephrasing the second sentence to something like: If unset, all JWT-SVIDs to which the workload is entitled are requested.

Author: I like it.

    string spiffe_id = 2;
}

// The JWTSVIDResponse message conveys JWT-SVIDs.
message JWTSVIDResponse {
    // Required. The list of returned JWT-SVIDs.
    repeated JWTSVID svids = 1;
}
```

### 5.2 Profile RPCs

#### 5.2.1 FetchX509SVID

> The `FetchX509SVID` RPC streams back X509-SVIDs, and X.509 bundles for authenticated X509-SVIDs for both the trust domain in which the server resides and foreign trust domains.

Reviewer 2: ```suggestion
The `FetchX509SVID` RPC streams back X509-SVIDs, and X.509 bundles for both the trust domain in which the server resides and foreign trust domains. These bundles MUST only be used to authenticate X.509-SVIDs.
```

Author: :+1:

    string audience = 1;

    // Required. The JWT-SVID to validate, encoded using JWS Compact
    // Serialization.
    string svid = 2;
}

// The ValidateJWTSVIDReponse message conveys the JWT-SVID validation results.
message ValidateJWTSVIDResponse {
    // Required. The SPIFFE ID of the validated JWT-SVID.
    string spiffe_id = 1;

    // Optional. Arbitrary claims contained within the payload of the validated
    // JWT-SVID.
    google.protobuf.Struct claims = 2;
}
```

### 6.2 Profile RPCs

#### 6.2.1 FetchJWTSVID

The `FetchJWTSVID` RPC allows clients to request one or more short-lived JWT-SVIDs for a specific audience.

The `JWTSVIDRequest` request message contains a mandatory `audience` field, which MUST contain the value to embed in the audience claim of the returned JWT-SVIDs. The `spiffe_id` field is optional, and is used to request a JWT-SVID for a specific SPIFFE ID. If unspecified, the server MUST return JWT-SVIDs for all identities authorized for the client. 

The `JWTSVIDResponse` response message consists of a mandatory `svids` field, which MUST contain one or more `JWTSVID` messages.

All fields in the `JWTSVID` message are mandatory.

> If the client is not authorized for any identities, nor the specific identity requested via the `spiffe_id` field, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information).

Reviewer 2: ```suggestion
If the client is not authorized for any identities, or not authorized for the specific identity requested via the `spiffe_id` field, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information).
```

Author: :+1:

#### 6.2.2 FetchJWTBundles

> The `FetchJWTBundles` RPC streams back JWT bundles for authenticated JWT-SVIDs for both the trust domain in which the server resides and foreign trust domains.

Reviewer 2: ```suggestion
The `FetchJWTBundles` RPC streams back JWT bundles for both the trust domain in which the server resides and foreign trust domains. These bundles MUST only be used to authenticate JWT-SVIDs.
```

Author: :+1:

    // X.509-SVID, its private key, and the bundle for the trust domain.
    repeated X509SVID svids = 1;

    // Optional. An ASN.1 DER encoded CRL.
    repeated bytes crl = 2;

    // Optional. CA certificate bundles belonging to foreign trust domains that the
    // workload should trust, keyed by the SPIFFE ID of the foreign
    // domain. Bundles are ASN.1 DER encoded.
    map<string, bytes> federated_bundles = 3;
}

// The X509SVID message carries a single SVID and all associated
// information, including X.509 bundle for the trust domain.
message X509SVID {
    // Required. The SPIFFE ID of the SVID in this entry
    string spiffe_id = 1;

    // Required. ASN.1 DER encoded certificate chain. MAY include
    // intermediates, the leaf certificate (or SVID itself) MUST come first.
    bytes x509_svid = 2;

    // Required. ASN.1 DER encoded PKCS#8 private key. MUST be unencrypted.
    bytes x509_svid_key = 3;

    // Required. ASN.1 DER encoded X.509 bundle for the trust domain.
    bytes bundle = 4;
}

// The X509BundlesRequest message conveys parameters for requesting X.509
// bundles. There are currently no such parameters.
message X509BundlesRequest {
}

> // The X509BundlesResponse message carries a global CRL and a

Reviewer: The `crl` field is still marked as repeated but the description here has dropped the notion of "a set of". This seems inconsistent.

Author: Good catch. I'll restore the original language and update the field descriptions for consistency.

As mentioned in [Stream Responses](#42-stream-responses), each `JWTBundleResponse` response contains the complete set of authorized JWT bundles for the client at that point in time. As such, if the server redacts bundles from a subsequent response (or all bundles, i.e., returns a "PermissionDenied" gRPC status code) the client SHOULD cease using the redacted bundles.

#### 6.2.3 ValidateJWTSVID

The `ValidateJWTSVID` RPC validates JWT-SVIDs for a specific audience on behalf of a client.

All fields in the `ValidateJWTSVIDRequest` and `ValidateJWTSVIDResponse` message are mandatory.

### 6.3 JWT-SVID Validation

Workload API clients SHOULD use the `ValidateJWTSVID` method for JWT validation if supported by the client, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.

> When interfacing with legacy JWT validators, the `FetchJWTBundles` method can be used to fetch JWKS bundles that can be used to validate JWT-SVID signatures. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API, it is possible to write a small shim that can retrieve the bundles and feed them to the legacy workload.

Reviewer: ```suggestion
When interfacing with legacy JWT validators, the `FetchJWTBundles` method can be used to fetch JWKS bundles that can be used to validate JWT-SVID signatures. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API (and thus cannot call `ValidateJWTSVID`), implementations can instead individually retrieve each bundle and feed them to the legacy workload for validation.
```

Author: :+1:

> The `FetchJWTBundles` method returns bundles keyed by the SPIFFE ID of the trust domain. When validating a JWT-SVID, the validator should use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.

Reviewer: ```suggestion
The `FetchJWTBundles` method returns bundles keyed by the SPIFFE ID of the trust domain. When validating a JWT-SVID, the validator MUST use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.
```

Author: Nice catch.

2\. [Extensibility](#2-extensibility)
3\. [Service Defintion](#3-service-definition)
4\. [Identifying the Caller](#4-identifying-the-caller)
5\. [X.509-SVID Profile](#5-x509-svid-profile)
5.1. [Workload API Client and Server Behavior](#51-workload-api-client-and-server-behavior)
5.2. [Federated Bundles](#52-federated-bundles)
5.3. [Default Identity](#53-default-identity)
5.4. [Profile Definition](#54-profile-definition)
5.5. [Default Values and Redacted Information](#55-default-values-and-redacted-information)
6\. [JWT-SVID Profile](#6-jwt-svid-profile)
6.1 [Client and Server Behavior](#61-client-and-server-behavior)
6.2 [Default Identity](#62-default-identity)
6.3 [Fetching Bundles](#63-fetching-bundles)
6.4 [Profile Definition](#64-profile-definition)
6.5. [Default Values and Redacted Information](#65-default-values-and-redacted-information)

## 1. Introduction

The SPIFFE Workload API is an API which provides information and services that enable workloads, or compute processes, to leverage SPIFFE identities and SPIFFE-based authentication systems. It is served by the [SPIFFE Workload Endpoint](SPIFFE_Workload_Endpoint.md), and comprises a number of services, or *profiles*.

> Currently, there are two profiles, both of which are mandatory:

Author: > both of which are mandatory

I couldn't remember where the group landed in terms of optionality. Happy to amend this if it isn't accurate.

Author: @evan2645 reminded me of what we discussed in SIG-SPEC a bit ago, that the profile is mandatory but can be administratively disabled.

### 6.1 Client and Server Behavior

The JWT-SVID Workload API profile exposes three gRPC methods: FetchJWTSVID, FetchJWTBundles, and ValidateJWTSVID.

The FetchJWTSVID and ValidateJWTSVID methods operate in a 1:1 request/response pattern. A single request made to the FetchJWTSVID method generates a single response containing the JWT-SVID(s) described in the request. Similarly, a single request made to the ValidateJWTSVID method generates a single response providing information about the token provided in the request. It should be noted that, under some implementations, the cost of creating a new connection may be high. Clients are encouraged to reuse connections when possible.

The FetchJWTBundles method is different. It is implemented as a gRPC server-side stream in order to facilitate rapid propagation of updates, namely key introductions or redactions. Every response message sent by the server MUST include the full set of information, and not just the information which has changed. This avoids complexity associated with state tracking on both Client and Server implementations.

> The client and server behavior of FetchJWTBundles is identical to that of the X509-SVID profile. Please see the Client and Server Behavior section of the X509-SVID profile for more detailed information.

Reviewer 2: I find this sentence a little hard to understand because you're comparing an RPC to a "profile" so it feels like apples and oranges to say they are identical.  The term behavior is also a little vague: clearly the RPCs needs to do some things that are different from one another, otherwise they would all be the same RPC, so I think we need to get a little more precise about what exactly is the same about them.

Author: Yes, I agree this is really unwieldy. In the larger refactoring I had started, there was only one section that described the streaming behavior, since what is mentioned in the X.509 profile is really just a property of all of the streaming RPCs in the Workload API. I'll see if I can't word smith this a little.

### 6.3 Fetching Bundles

The JWT-SVID Workload API profile exposes a method for fetching JWKS bundles that can be used to validate JWT-SVID signatures. This method is exposed for the purpose of supporting legacy JWT validators. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API, it is possible to write a small shim that can retrieve the bundles and feed them to the legacy workload.

JWT-SVID signing keys may represent only a subset of the keys present in a SPIFFE trust bundle. Implementers of the SPIFFE Workload API MUST NOT include keys with other uses in the returned JWKS bundles. In other words, the SPIFFE Workload API should only provide JWT-SVID validators with bundle members that are valid JWT-SVID signers.

The JWTBundlesResponse message includes a map of JWKS bundles, keyed by trust domain. When validating a JWT-SVID, the validator should use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.

> Please note that nominally, workloads will use the ValidateJWTSVID method for JWT validation, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.

Reviewer 2: ```suggestion
Workload API clients SHOULD use the ValidateJWTSVID method for JWT validation if supported by the client, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.
```

Author: :+1:
